llvm/include/llvm/ADT/PagedVector.h

```diff
@@ -0,0 +1,301 @@
+  /// resize method.
+  [[nodiscard]] size_t size() const { return Size; }
+
+  /// Resize the vector. Notice that the constructor of the elements will not
+  /// be invoked until an element of a given page is accessed, at which point
+  /// all the elements of the page will be constructed.
+  ///
+  /// If the new size is smaller than the current size, the elements of the
+  /// pages that are not needed anymore will be destroyed, however, elements of
+  /// the last page will not be destroyed.
+  ///
+  /// For these reason the usage of this vector is discouraged if you rely
+  /// on the construction / destructor of the elements to be invoked.
+  void resize(size_t NewSize) {
+    if (NewSize == 0) {
+      clear();
+      return;
+    }
+    // Handle shrink case: destroy the elements in the pages that are not
+    // needed anymore and deallocate the pages.
+    //
+    // On the other hand, we do not destroy the extra elements in the last page,
+    // because we might need them later and the logic is simpler if we do not
+    // destroy them. This means that elements are only destroyed only when the
+    // page they belong to is destroyed. This is similar to what happens on
+    // access of the elements of a page, where all the elements of the page are
+    // constructed not only the one effectively neeeded.
+    size_t NewLastPage = (NewSize - 1) / PageSize;
+    if (NewSize < Size) {
+      // Destruct the elements in the pages that are not needed anymore.
+      // Notice that we need to do this only if the constructor of the elements
+      // is not trivial.
+      if constexpr (!std::is_trivially_destructible_v<T>) {
+        for (size_t I = NewLastPage + 1, N = PageToDataPtrs.size(); I < N; ++I) {
+          T *Page = PageToDataPtrs[I];
+          // We need to invoke the destructor on all the elements of the page.
+          if (Page != InvalidPage)
+            std::destroy_n(Page, PageSize);
+        }
+      }
+      for (size_t I = NewLastPage + 1, N = PageToDataPtrs.size(); I < N; ++I) {
+        T *Page = PageToDataPtrs[I];
+        if (Page != InvalidPage)
+          Allocator.getPointer()->Deallocate(Page);
+        // We mark the page invalid, to avoid double deletion.
+        PageToDataPtrs[I] = InvalidPage;
+      }
+      PageToDataPtrs.resize(NewLastPage + 1);
+    }
+    Size = NewSize;
+    // If the capacity is enough, just update the size and continue
+    // with the currently allocated pages. Notice that we do not
+    // need to default construct any new element, because that was already done
+    // when the page was allocated.
+    if (Size <= capacity())
+      return;
+    // We use InvalidPage to indicate that a page has not been allocated yet.
+    // This cannot be 0, because 0 is a valid page id.
+    // We use InvalidPage instead of a separate bool to avoid wasting space.
+    assert(NewLastPage + 1 > PageToDataPtrs.size());
+    PageToDataPtrs.resize(NewLastPage + 1, InvalidPage);
```

Comment:
The end of this function is unnecessarily distinguishing between the case of growing and shrinking, and can be simplified a bit:

```suggestion
    }

    Size = NewSize;
    PageToDataPtrs.resize(NewLastPage + 1, InvalidPage);
```